diff --git a/src/local.mk b/src/local.mk
index 13eeea8e0..6cc5ac302 100644
--- a/src/local.mk
+++ b/src/local.mk
@@ -164,7 +164,7 @@ src_numfmt_LDADD = $(LDADD)
 src_od_LDADD = $(LDADD)
 src_paste_LDADD = $(LDADD)
 src_pathchk_LDADD = $(LDADD)
-src_pinky_LDADD = $(LDADD)
+src_pinky_LDADD = $(LDADD) -lsystemd
 src_pr_LDADD = $(LDADD)
 src_printenv_LDADD = $(LDADD)
 src_printf_LDADD = $(LDADD)
@@ -206,11 +206,11 @@ src_uname_LDADD = $(LDADD)
 src_unexpand_LDADD = $(LDADD)
 src_uniq_LDADD = $(LDADD)
 src_unlink_LDADD = $(LDADD)
-src_uptime_LDADD = $(LDADD)
-src_users_LDADD = $(LDADD)
+src_uptime_LDADD = $(LDADD) -lsystemd
+src_users_LDADD = $(LDADD) -lsystemd
 # See vdir_LDADD below
 src_wc_LDADD = $(LDADD)
-src_who_LDADD = $(LDADD)
+src_who_LDADD = $(LDADD) -lsystemd
 src_whoami_LDADD = $(LDADD)
 src_yes_LDADD = $(LDADD)
 
