diff --git a/src/uptime.c b/src/uptime.c
index 061b85c16..a970510b0 100644
--- a/src/uptime.c
+++ b/src/uptime.c
@@ -22,6 +22,10 @@
 #include <sys/types.h>
 #include "system.h"
 
+#if defined(WITH_SYSTEMD) || defined(WITH_ELOGIND)
+#include <systemd/sd-login.h>
+#endif
+
 #if HAVE_SYSCTL && HAVE_SYS_SYSCTL_H && ! defined __GLIBC__
 # include <sys/sysctl.h>
 #endif
@@ -102,6 +106,12 @@ print_uptime (size_t n, const STRUCT_UTMP *this)
   }
 #endif
 
+#if defined(WITH_SYSTEMD) || defined(WITH_ELOGIND)
+  if (this == NULL)
+    entries = sd_get_sessions (NULL);
+  else
+    {
+#endif
 #if HAVE_STRUCT_UTMP_UT_TYPE || HAVE_STRUCT_UTMPX_UT_TYPE
   /* Loop through all the utmp entries we just read and count up the valid
      ones, also in the process possibly gleaning boottime. */
@@ -116,6 +126,9 @@ print_uptime (size_t n, const STRUCT_UTMP *this)
   (void) n;
   (void) this;
 #endif
+#if defined(WITH_SYSTEMD) || defined(WITH_ELOGIND)
+    }
+#endif
 
   time_now = time (NULL);
 #if defined HAVE_PROC_UPTIME
@@ -241,7 +254,11 @@ main (int argc, char **argv)
   switch (argc - optind)
     {
     case 0:			/* uptime */
+#if defined(WITH_SYSTEMD) || defined(WITH_ELOGIND)
+      print_uptime (0, NULL);
+#else
       uptime (UTMP_FILE, READ_UTMP_CHECK_PIDS);
+#endif
       break;
 
     case 1:			/* uptime <utmp file> */
